tasks_management/services.py

```diff
@@ -1,16 +1,78 @@
 import logging
-from typing import Type, Dict
+from datetime import date
+from typing import Dict, Type
 
 from django.db import transaction
 
-from core.services import create_or_update_core_user, BaseService
-from core.services.utils import check_authentication, output_exception, model_representation, output_result_success
-from tasks_management.models import TaskGroup, TaskExecutor
-from tasks_management.validation import TaskGroupValidation, TaskExecutorValidation
+from core.services import BaseService
+from core.signals import register_service_signal
+from core.services.utils import check_authentication, output_exception
+
+from tasks_management.models import TaskGroup, TaskExecutor, Task
+from tasks_management.validation import TaskGroupValidation, TaskExecutorValidation, TaskValidation
 
 logger = logging.getLogger(__name__)
 
 
+class TaskService(BaseService):
+    OBJECT_TYPE = Task
+
+    def __init__(self, user, validation_class=TaskValidation):
+        super().__init__(user, validation_class)
+
+    @register_service_signal('task_service.create')
+    def create(self, obj_data):
+        return super().create(obj_data)
+
+    @register_service_signal('task_service.update')
+    def update(self, obj_data):
+        return super().update(obj_data)
+
+    @register_service_signal('task_service.delete')
+    def delete(self, obj_data):
+        return super().delete(obj_data)
+
+    @register_service_signal('task_service.execute_task')
+    def execute_task(self, obj_data):
+        obj = self.OBJECT_TYPE.objects.filter(id=obj_data['id']).first()
+        if not obj:
+            return self._result(False, None, 'Task not found', str(obj_data['id']))
+
+        return self._result(True, data={'task': obj})
+
+    @register_service_signal('task_service.complete_task')
+    def complete_task(self, obj_data):
+        obj = self.OBJECT_TYPE.objects.filter(id=obj_data['id']).first()
+        if not obj:
+            return self._result(False, None, 'Task not found', str(obj_data['id']))
+
+        obj.status = Task.Status.FAILED if obj_data.get('failed', False) else Task.Status.COMPLETED
+        obj.save(username=self.user.login_name)
+        return self._result(True, data={'task': obj})
+
+    def _base_payload_adjust(self, obj_data):
+        data = obj_data.pop('data', {})
+        if isinstance(data, dict):
+            self._convert_dates_to_strings(data)
+        return {**obj_data, "data": data}
+
+    def _convert_dates_to_strings(self, dictionary):
+        for key, value in dictionary.items():
+            if isinstance(value, date):
+                dictionary[key] = value.isoformat()
+
+    def _result(self, success, data=None, message='', details=''):
```

Comment:
```
def _result(self, success, data=None, message='', details=''):
    data = {} if data is None else data
    result = {
        'success': success,
        **({'data': data} if success else {'message': message, 'details': details})
    }
    return result
```
Please tell me if such refactored method could work.

Reply:
We could make it even shorter:
```python
def _result(self, success, data=None, message='', details=''):
    return {
        'success': success,
        **({'data': data or {}} if success else {'message': message, 'details': details})
    }
```
while it's good to reduce number of lines I think that `**({'data': data} if success else {'message': message, 'details': details})` is not that readable and less straightforward. 

I think another approach for this _result could be: 
```python
def _result(self, success, data=None, message='', details=''):
        return {
           'data': data or {}, 
           'message': message if not success, 
           'details':  details if not success, 
        }
```

@malinowskikam please let me know if this set of keys is standard across the Services in application. I think that using conditional 'message' and 'details' in case of success=False is quite odd. I'd like to know if we could wrap it in general `error': {'message': message,  'details':  details}`

Reply:
@dborowiecki Yes, this dict format is a standard for all services in the system, but the core solution adds `message` and `details` keys even if `success=True`.

@dborowiecki @sniedzielski I changed custom implementation of `_result` with the one from core. I also updated the error handling to match the one from core service.